File: labs-algo/1/K.py

#
# @author Saveliy Bakturin
#
# Don't write off, if you don't wanna be banned!
#

class Stack:
    def __init__(self):
        self.items = []

    def empty(self):
        return self.items == []

    def push(self, item):
        self.items.append(item)

    def pop(self):
        return self.items.pop()

    def peek(self):
        return self.items[len(self.items) - 1]

    def size(self):
        return len(self.items)


n = int(input())
answer, counter, leftStackValue, rightStackValue, leftStackMin, rightStackMin = [], 0, Stack(), Stack(), Stack(), Stack()
for i in range(n):
    inputList = input().split()
    if inputList[0] == "+":
        temp = int(inputList[1])
        if leftStackValue.empty():
            temp = temp
        else:
            temp = min(temp, leftStackMin.peek())
        leftStackValue.push(int(inputList[1]))
        leftStackMin.push(temp)
        counter += 1
        if not(rightStackValue.empty()) and not(leftStackValue.empty()):
            temp = min(leftStackMin.peek(), rightStackMin.peek())
        else:
            if rightStackValue.empty():
                temp = leftStackMin.peek()
            else:
                temp = rightStackMin.peek()
        answer.append(temp)
    else:
        if counter == 0:
            answer.append(-1)
        else:
            if rightStackValue.empty():
                while not(leftStackValue.empty()):
                    temp_min = 0
                    temp_value = leftStackValue.peek()
                    leftStackValue.pop()
                    if rightStackValue.empty():
                        temp_min = temp_value
                    else:
                        temp_min = min(temp_value, rightStackMin.peek())
                    rightStackValue.push(temp_value)
                    rightStackMin.push(temp_min)
            rightStackValue.pop()
            rightStackMin.pop()
            counter -= 1
            if counter == 0:
                answer.append(-1)
            else:
                if not(rightStackValue.empty()) and not(leftStackValue.empty()):
                    temp = min(leftStackMin.peek(), rightStackMin.peek())
                else:
                    if rightStackValue.empty():
                        temp = leftStackMin.peek()
                    else:
                        temp = rightStackMin.peek()
                answer.append(temp)
for i in answer:
    print(i)

File: algo/1/K.cpp
#include <cstdio>
#include <stack>

/**
 * @author Saveliy Bakturin
 * <p>
 * Don't write off, if you don't wanna be banned!
 */

class Queue {
private:
	std::stack<int> left, right, leftMin, rightMin;

	void print();
public:
	Queue() = default;
	~Queue() = default;

	void add(const int x);
	void remove();
};

void Queue::add(const int x) {
	left.push(x);
	if (leftMin.empty()) {
		leftMin.push(x);
	} else {
		leftMin.push(x > leftMin.top() ? leftMin.top() : x);
	}
	Queue::print();
};

void Queue::remove() {
	if (right.empty()) {
		while (!left.empty()) {
			int x = left.top();
			left.pop();
			leftMin.pop();
			right.push(x);
			if (rightMin.empty()) {
				rightMin.push(x);
			} else {
				rightMin.push(x > rightMin.top() ? rightMin.top() : x);
			}
		}
	}
	right.pop();
	rightMin.pop();
	Queue::print();
};

void Queue::print() {
	if (!leftMin.empty() && !rightMin.empty()) {
		printf("%i\n", std::min(leftMin.top(), rightMin.top()));
	} else {
		if (leftMin.empty() && !rightMin.empty()) {
			printf("%i\n", rightMin.top());
		} else {
			if (!leftMin.empty() && rightMin.empty()) {
				printf("%i\n", leftMin.top());
			} else {
				printf("-1\n");
			}
		}
	}
};

int main(void) {
	int q;
	scanf("%i", &q);
	Queue queue;
	for (int i = 0; i < q; i++) {
		char op;
		scanf(" %c", &op);
		if (op == '+') {
			int val;
			scanf("%i", &val);
			queue.add(val);
		} else {
			queue.remove();
		}
	}
}
